File: BOJ/java/10971.java
/*
 * 외판원 순회 2
 * https://www.acmicpc.net/problem/10971
 */

import java.util.Scanner;

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[][] a = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                a[i][j] = sc.nextInt();
            }
        }

        int[] d = new int[n];
        for (int i = 0; i < n; i++) d[i] = i;

        int result = Integer.MAX_VALUE;

        do {
            boolean ok = true;
            int sum = 0;
            for (int i = 0; i < n - 1; i++) {
                if (a[d[i]][d[i + 1]] == 0) {
                    ok = false;
                } else {
                    sum += a[d[i]][d[i + 1]];
                }
            }
            if (ok && a[d[n - 1]][d[0]] != 0) {
                sum += a[d[n - 1]][d[0]];
                if (result > sum) {
                    result = sum;
                }
            }
        } while (nextPermutation(d));
        System.out.println(result);
    }

    static boolean nextPermutation(int[] a) {
        int lastIndex = a.length - 1;
        int i = lastIndex;
        while (i > 0 && a[i] < a[i - 1]) {
            i--;
        }
        if (i == 0) return false;

        int j = lastIndex;
        while (j >= i && a[j] < a[i - 1]) {
            j--;
        }
        swap(a, i - 1, j);

        j = lastIndex;
        while (i < j) swap(a, i++, j--);
        return true;
    }

    static void swap(int[] a, int i, int j) {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }
}

File: BOJ/kotlin/10971.kt
fun main() {
    val n = readLine()!!.toInt()
    val a = Array(n) { IntArray(n) { 0 } }
    repeat(n) { i -> a[i] = readLine()!!.split(" ").map { it.toInt() }.toIntArray() }
    val d = (0 until n).toMutableList()

    var result = Int.MAX_VALUE

    do {
        var ok = true
        var sum = 0
        for (i in 0 until n - 1) {
            if (a[d[i]][d[i + 1]] == 0) ok = false
            else sum += a[d[i]][d[i + 1]]
        }

        if (ok && a[d[n - 1]][d[0]] != 0) {
            sum += a[d[n - 1]][d[0]]
            if (result > sum) result = sum
        }
    } while (nextPermutation(d))
    println(result)
}

fun nextPermutation(a: MutableList<Int>): Boolean {
    val lastIndex = a.size - 1
    var i = (1..lastIndex).findLast { a[it] > a[it - 1] } ?: return false
    var j = (i..lastIndex).findLast { a[it] > a[i - 1] } ?: return false
    a.swap(i - 1, j)
    j = lastIndex
    while (i < j) a.swap(i++, j--)
    return true
}

fun <T> MutableList<T>.swap(i: Int, j: Int) {
    val temp = this[i]
    this[i] = this[j]
    this[j] = temp
}